### PR TITLE
Feature/add gRPC

### PR DESCRIPTION
**Pull request by LoneXDii**

### CODE REVIEW

File: OrderService/OrderService.Application/UseCases/CartUseCases/AddProductToCartUseCase/AddProductToCartRequestHandler.cs [hunk 1]
```diff
@@ -9,18 +9,25 @@ internal class AddProductToCartRequestHandler(ITemporaryStorageService temporary
...
     {
-        var product = await productService.GetByIdIfSufficientQuantityAsync(request.product.Id, request.product.Quantity);
+        var cartProductQuantity = 0;
+
+        var cart = await temporaryStorage.GetCartAsync(cancellationToken);
+
+        if (cart.ContainsKey(request.product.Id))
+        {
+            cartProductQuantity = cart[request.product.Id].Quantity;
+        }
```
Comment (UzheNeKate) [inline OrderService/OrderService.Application/UseCases/CartUseCases/AddProductToCartUseCase/AddProductToCartRequestHandler.cs]: what do you think about moving this to a separate method which returns result ProductQuantity?

File: ProductsService/ProductsService.API/Services/ProductsGrpcService.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    [Authorize]
+    public override async Task<ProductsListResponse> TakeProducts(ProductsListRequest request, ServerCallContext context)
+    {
+        var dict = request.Products.ToDictionary(pr => pr.Id, pr => pr.Quantity);
+
+        var data = await _mediator.Send(new TakeProductsRequest(dict));
+
+        var products = _mapper.Map<List<ProductResponse>>(data);
+
+        var response = new ProductsListResponse();
```
Comment (UzheNeKate) [inline ProductsService/ProductsService.API/Services/ProductsGrpcService.cs]: maybe let's pass products as a constructor parameter?

File: ProductsService/ProductsService.Application/UseCases/ProductUseCases/Queries/TakeProducts/TakeProductsRequestHandler.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+            product.Quantity -= quantity;
+
+            await unitOfWork.ProductCommandRepository.UpdateAsync(product, cancellationToken);
+        }
+
+        if (errors.Length > 0)
+        {
+            throw new BadRequestException(errors.ToString());
+        }
```
Comment (UzheNeKate) [inline ProductsService/ProductsService.Application/UseCases/ProductUseCases/Queries/TakeProducts/TakeProductsRequestHandler.cs]: please move this logic to a separate method

**Review by UzheNeKate — CHANGES_REQUESTED** [review]
please refactor code

File: ProductsService/ProductsService.API/Services/ProductsGrpcService.cs (hunk 2 above)
Comment (LoneXDii) [reply ProductsService/ProductsService.API/Services/ProductsGrpcService.cs]: ProductsListResponse is an autogenerated class from a .proto file. It only has an empty constructor and a constructor that takes another ProductsListResponse as a parameter

Comment (LoneXDii) [reply ProductsService/ProductsService.API/Services/ProductsGrpcService.cs]: I've created a mapping profile from List of ProductResponse to ProductsListResponse

**Review by UzheNeKate — APPROVED** [review]

**Pull request merged**